### PR TITLE
 Add FileCheck annotations to MIR-opt SROA tests

### PR DESCRIPTION
**Pull request by JarlEvanson**
Part of #116971, adds FileCheck annotations to SROA MIR-opt tests in `tests/mir-opt/sroa` and a few uncategorized files.

r? cjgillot

### CODE REVIEW

File: tests/mir-opt/sroa/lifetimes.rs [hunk 1]
```diff
@@ -16,6 +15,9 @@ struct Foo<T: Err> {
 
 // EMIT_MIR lifetimes.foo.ScalarReplacementOfAggregates.diff
 fn foo<T: Err>() {
+    // CHECK-LABEL: fn foo
```
Comment (cjgillot) [inline tests/mir-opt/sroa/lifetimes.rs:18]: As said in the comment, this test is about verifying that we erase the lifetimes in the types of locals. Could you add the annotations for this?

File: tests/mir-opt/sroa/structs.rs [hunk 2]
```diff
@@ -13,28 +12,57 @@ impl Drop for Tag {
...
 }
 
+/// Check that SROA excludes structs with a `Drop` implementation.
 pub fn dropping() {
+    // CHECK-LABEL: fn dropping(
+
+    // CHECK: [[aggregate:_[0-9]+]]: S;
+
+    // CHECK: bb0: {
+    // CHECK-NOT: scope
```
Comment (cjgillot) [inline tests/mir-opt/sroa/structs.rs:22]: Why this line?

File: tests/mir-opt/sroa/structs.rs [hunk 3]
```diff
@@ -13,28 +12,57 @@ impl Drop for Tag {
...
+    // CHECK-LABEL: fn structs(
     struct U {
         _foo: usize,
         a: f32,
     }
+    // CHECK: [[struct:_[0-9]+]]: structs::U;
 
+    // CHECK: bb0: {
+    // CHECK-NOT: scope
+    // CHECK-NOT: [[struct]]
```
Comment (cjgillot) [inline tests/mir-opt/sroa/structs.rs:50]: Could you add checks that the decomposed assignments are correct?

File: tests/mir-opt/sroa/structs.rs [hunk 4]
```diff
@@ -46,11 +74,16 @@ struct Foo {
...
-
+/// Check that non-escaping uses of a struct are destructured.
 pub fn flat() {
+    // CHECK-LABEL: fn flat(
+
+    // CHECK: [[struct:_[0-9]+]]: Foo;
+
+    // CHECK: bb0: {
+    // CHECK-NOT: scope
+    // CHECK-NOT: [[struct]]
```
Comment (cjgillot) [inline tests/mir-opt/sroa/structs.rs:85]: Could you add checks that the decomposed assignments are correct?

File: tests/mir-opt/sroa/structs.rs [hunk 5]
```diff
@@ -46,11 +74,16 @@ struct Foo {
...
+    // CHECK: fn g
```
Comment (cjgillot) [inline tests/mir-opt/sroa/structs.rs:86]: This is not meant to appear here.

File: tests/mir-opt/sroa/structs.rs [hunk 6]
```diff
@@ -76,31 +113,72 @@ fn f(a: *const u32) {
...
+    // CHECK-LABEL: fn copies(
+
+    // CHECK: [[external:_[0-9]+]]: Foo) ->
+    // CHECK: [[internal0:_[0-9]+]]: Foo;
+    // CHECK: [[internal1:_[0-9]+]]: Foo;
+
+    // CHECK: bb0: {
+    // CHECK-NOT: scope
+    // CHECK-NOT: [[internal0]]
+    // CHECK-NOT: [[internal1]]
```
Comment (cjgillot) [inline tests/mir-opt/sroa/structs.rs:138]: Could you check that the assignments are correct?

File: tests/mir-opt/sroa/structs.rs [hunk 7]
```diff
@@ -76,31 +113,72 @@ fn f(a: *const u32) {
...
+/// the original struct.
 fn ref_copies(x: &Foo) {
+    // CHECK-LABEL: fn ref_copies(
+
+    // CHECK: [[external:_[0-9]+]]: &Foo) ->
+    // CHECK: [[internal:_[0-9]+]]: Foo;
+
+    // CHECK: bb0: {
+    // CHECK-NOT: scope
+    // CHECK-NOT: [[internal]]
```
Comment (cjgillot) [inline tests/mir-opt/sroa/structs.rs:156]: Could you check that the assignments are correct?

File: tests/mir-opt/sroa/structs.rs [hunk 8]
```diff
@@ -76,31 +113,72 @@ fn f(a: *const u32) {
...
+/// reassigned to copy from the constant.
 fn constant() {
+    // CHECK-LABEL: constant(
+
+    // CHECK: [[constant:_[0-9]+]]: (usize, u8);
+
+    // CHECK: bb0: {
+    // CHECK-NOT: scope
+    // CHECK-NOT: [[constant]]
+    // CHECK: [[constant]] = const
```
Comment (cjgillot) [inline tests/mir-opt/sroa/structs.rs:173]: Could you check that the assignments are correct?

**Comment by JarlEvanson** [comment]
Thanks for the review. I think I implemented all your suggestions, but let me know if I didn't.

File: tests/mir-opt/remove_storage_markers.rs [hunk 9]
```diff
@@ -8,6 +7,10 @@
 
 // EMIT_MIR remove_storage_markers.main.RemoveStorageMarkers.diff
 fn main() {
+    // CHECK-LABLE: fn main(
```
Comment (cjgillot) [inline tests/mir-opt/remove_storage_markers.rs:10]: Typo
```suggestion
    // CHECK-LABEL: fn main(
```

File: tests/mir-opt/sroa/structs.rs [hunk 10]
```diff
@@ -13,28 +12,65 @@ impl Drop for Tag {
...
+
+    // CHECK: bb0: {
+    // CHECK: [[enum]] = Option::<usize>::Some
     if let Some(a) = Some(a) { a } else { 0 }
 }
 
+/// Check that SROA destructures `U`.
 pub fn structs(a: f32) -> f32 {
+    // CHECK-LABEL: fn structs(
+    // CHECK-SAME: [[a:_[0-9]+]]: f32
```
Comment (cjgillot) [inline tests/mir-opt/sroa/structs.rs:40]: `a` will always be `_1`. No need for a pattern here.

File: tests/mir-opt/sroa/structs.rs [hunk 11]
```diff
@@ -13,28 +12,65 @@ impl Drop for Tag {
...
         _foo: usize,
         a: f32,
     }
-
+    // CHECK: [[ret:_0]]: f32;
+    // CHECK-DAG: [[struct:_[0-9]+]]: structs::U;
+    // CHECK-DAG: [[foo:_[0-9]+]]: usize;
+    // CHECK-DAG: [[a_tmp:_[0-9]+]]: f32;
+    // CHECK-NOT: Stop reordering
+    // CHECK-DAG: [[a_ret:_[0-9]+]]: f32;
```
Comment (cjgillot) [inline tests/mir-opt/sroa/structs.rs:50]: Could you use simple `CHECK`s? The order should be deterministic. (This comment applies everywhere you use CHECK-DAG).

File: tests/mir-opt/sroa/structs.rs [hunk 12]
```diff
@@ -13,28 +12,65 @@ impl Drop for Tag {
...
+
+    // CHECK: bb0: {
+    // CHECK: [[a_tmp]] = [[a]];
+    // CHECK-NOT: [[struct]]
+    // CHECK-DAG: [[foo]] = const 0_usize;
+    // CHECK-DAG: [[a_ret]] = move [[a_tmp]];
```
Comment (cjgillot) [inline tests/mir-opt/sroa/structs.rs:56]: Likewise.

File: tests/mir-opt/sroa/structs.rs [hunk 13]
```diff
@@ -13,28 +12,65 @@ impl Drop for Tag {
...
 pub fn unions(a: f32) -> u32 {
+    // CHECK-LABEL: fn unions(
     union Repr {
         f: f32,
         u: u32,
     }
+    // CHECK: [[union:_[0-9]+]]: unions::Repr;
+
+    // CHECK: bb0: {
+    // CHECK-NOT: scope
```
Comment (cjgillot) [inline tests/mir-opt/sroa/structs.rs:72]:
```suggestion
```

File: tests/mir-opt/sroa/structs.rs [hunk 14]
```diff
@@ -13,28 +12,65 @@ impl Drop for Tag {
...
+    // CHECK: [[union]] = Repr {
```
Comment (cjgillot) [inline tests/mir-opt/sroa/structs.rs:73]: Should we add a `_0 = ([[union]].1: u32)`?

File: tests/mir-opt/sroa/structs.rs [hunk 15]
```diff
@@ -13,28 +12,65 @@ impl Drop for Tag {
...
 }
 
+/// Check that SROA excludes enums.
 pub fn enums(a: usize) -> usize {
+    // CHECK-LABEL: fn enums(
+
+    // CHECK: [[enum:_[0-9]+]]: std::option::Option<usize>;
+
+    // CHECK: bb0: {
+    // CHECK: [[enum]] = Option::<usize>::Some
```
Comment (cjgillot) [inline tests/mir-opt/sroa/structs.rs:33]: Should we add a `CHECK: _0 = (([[enum]] as Some).0: usize);`?

File: tests/mir-opt/sroa/structs.rs [hunk 16]
```diff
@@ -76,31 +128,88 @@ fn f(a: *const u32) {
...
+/// Check that deaggregated assignments from constants are placed after the constant's
+/// assignment. Also check that copying field accesses from the copy of the constant are
+/// reassigned to copy from the constant.
 fn constant() {
+    // CHECK-LABEL: constant(
+
+    // CHECK: [[constant:_[0-9]+]]: (usize, u8);
+
+    // CHECK: bb0: {
+    // CHECK-NOT: scope
```
Comment (cjgillot) [inline tests/mir-opt/sroa/structs.rs:202]:
```suggestion
```

File: tests/mir-opt/sroa/structs.rs [hunk 17]
```diff
@@ -76,31 +128,88 @@ fn f(a: *const u32) {
...
+    // CHECK-NOT: [[constant]]
+    // CHECK: [[constant]] = const
     const U: (usize, u8) = (5, 9);
     let y = U;
     let t = y.0;
     let u = y.1;
```
Comment (cjgillot) [inline tests/mir-opt/sroa/structs.rs:208]: Could you add the patterns for the assignments from `[[constant]]` to those locals?

**Review by cjgillot — COMMENTED** [review]
Thanks @JarlEvanson. A few more nits, and will be good to go.

File: tests/mir-opt/sroa/structs.rs (hunk 17 above)
Comment (JarlEvanson) [reply tests/mir-opt/sroa/structs.rs:212]: done

**Comment by cjgillot** [comment]
Thanks!
@bors r+ rollup

**Pull request merged**